File: services/report-service/src/main/java/com/otterworks/report/controller/ReportController.java
package com.otterworks.report.controller;

import com.otterworks.report.model.Report;
import com.otterworks.report.model.ReportRequest;
import com.otterworks.report.model.ReportResponse;
import com.otterworks.report.model.ReportStatus;
import com.otterworks.report.service.ReportService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import jakarta.validation.Valid;
import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * REST controller for report management.
 */
@RestController
@RequestMapping("/api/v1/reports")
@Tag(name = "Reports", description = "Report generation and management")
public class ReportController {

    private static final Logger logger = LoggerFactory.getLogger(ReportController.class);

    private final ReportService reportService;

    public ReportController(ReportService reportService) {
        this.reportService = reportService;
    }

    @PostMapping
    @Operation(summary = "Create a new report", description = "Submits a report generation request. The report is generated asynchronously.")
    @ApiResponses({
            @ApiResponse(responseCode = "202", description = "Report request accepted"),
            @ApiResponse(responseCode = "400", description = "Invalid request")
    })
    public ResponseEntity<ReportResponse> createReport(
            @Valid @RequestBody ReportRequest request) {

        logger.info("Report request: name={}, category={}, type={}, by={}",
                request.getReportName(), request.getCategory(),
                request.getReportType(), request.getRequestedBy());

        Report report = reportService.createReport(request);
        return ResponseEntity.status(HttpStatus.ACCEPTED)
                .body(ReportResponse.fromEntity(report));
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get report by ID", description = "Returns the report metadata and status")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Report found"),
            @ApiResponse(responseCode = "404", description = "Report not found")
    })
    public ResponseEntity<ReportResponse> getReport(
            @Parameter(description = "Report ID", required = true)
            @PathVariable Long id) {

        Optional<Report> report = reportService.getReport(id);
        if (!report.isPresent()) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(ReportResponse.fromEntity(report.get()));
    }

    @GetMapping
    @Operation(summary = "List reports", description = "List reports filtered by user ID or status")
    public ResponseEntity<Map<String, Object>> listReports(
            @Parameter(description = "Filter by user ID")
            @RequestParam(required = false) String userId,
            @Parameter(description = "Filter by status")
            @RequestParam(required = false) ReportStatus status) {

        List<Report> reports;
        if (userId != null) {
            reports = reportService.getReportsByUser(userId);
        } else if (status != null) {
            reports = reportService.getReportsByStatus(status);
        } else {
            reports = reportService.getReportsByStatus(ReportStatus.COMPLETED);
        }

        List<ReportResponse> responses = reports.stream()
                .map(ReportResponse::fromEntity)
                .collect(Collectors.toList());

        Map<String, Object> response = new HashMap<>();
        response.put("reports", responses);
        response.put("total", responses.size());

        return ResponseEntity.ok(response);
    }

    @GetMapping("/{id}/download")
    @Operation(summary = "Download a generated report", description = "Returns the report file for download")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Report file"),
            @ApiResponse(responseCode = "404", description = "Report not found or not yet completed"),
            @ApiResponse(responseCode = "409", description = "Report is still generating")
    })
    public ResponseEntity<Resource> downloadReport(
            @Parameter(description = "Report ID", required = true)
            @PathVariable Long id) {

        Optional<Report> optReport = reportService.getReport(id);
        if (!optReport.isPresent()) {
            return ResponseEntity.notFound().build();
        }

        Report report = optReport.get();

        if (report.getStatus() == ReportStatus.GENERATING || report.getStatus() == ReportStatus.PENDING) {
            return ResponseEntity.status(HttpStatus.CONFLICT).build();
        }

        if (report.getStatus() == ReportStatus.FAILED || report.getFilePath() == null) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }

        File file = new File(report.getFilePath());
        if (!file.exists()) {
            logger.error("Report file missing: {}", report.getFilePath());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }

        try {
            byte[] fileContent = FileUtils.readFileToByteArray(file);
            ByteArrayResource resource = new ByteArrayResource(fileContent);

            String contentType = getContentType(report.getReportType());
            String fileName = file.getName();

            return ResponseEntity.ok()
                    .contentType(MediaType.parseMediaType(contentType))
                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + fileName + "\"")
                    .contentLength(fileContent.length)
                    .body(resource);

        } catch (IOException e) {
            logger.error("Failed to read report file {}: {}", report.getFilePath(), e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "Delete a report", description = "Deletes the report record and its generated file")
    @ApiResponses({
            @ApiResponse(responseCode = "204", description = "Report deleted"),
            @ApiResponse(responseCode = "404", description = "Report not found")
    })
    public ResponseEntity<Void> deleteReport(
            @Parameter(description = "Report ID", required = true)
            @PathVariable Long id) {

        boolean deleted = reportService.deleteReport(id);
        if (!deleted) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.noContent().build();
    }

    // ----- Private helpers -----

    private String getContentType(com.otterworks.report.model.ReportType reportType) {
        switch (reportType) {
            case PDF:
                return "application/pdf";
            case CSV:
                return "text/csv";
            case EXCEL:
                return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
            default:
                return "application/octet-stream";
        }
    }
}

File: services/report-service/Controllers/ReportController.cs
using FluentValidation;
using Microsoft.AspNetCore.Mvc;
using OtterWorks.ReportService.Models;
using OtterWorks.ReportService.Services;

namespace OtterWorks.ReportService.Controllers;

[ApiController]
[Route("api/v1/reports")]
public class ReportController : ControllerBase
{
    private readonly IReportService _reportService;
    private readonly IValidator<ReportRequest> _validator;
    private readonly ILogger<ReportController> _logger;

    public ReportController(
        IReportService reportService,
        IValidator<ReportRequest> validator,
        ILogger<ReportController> logger)
    {
        _reportService = reportService;
        _validator = validator;
        _logger = logger;
    }

    [HttpPost]
    [ProducesResponseType(typeof(ReportResponse), StatusCodes.Status202Accepted)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> CreateReport([FromBody] ReportRequest request)
    {
        var validation = await _validator.ValidateAsync(request);
        if (!validation.IsValid)
        {
            return BadRequest(new { errors = validation.Errors.Select(e => e.ErrorMessage) });
        }

        _logger.LogInformation("Report request: name={Name}, category={Category}, type={Type}, by={By}",
            request.ReportName, request.Category, request.ReportType, request.RequestedBy);

        var report = await _reportService.CreateReportAsync(request);
        return StatusCode(StatusCodes.Status202Accepted, ReportResponse.FromEntity(report));
    }

    [HttpGet("{id:long}")]
    [ProducesResponseType(typeof(ReportResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetReport(long id)
    {
        var report = await _reportService.GetReportAsync(id);
        if (report == null)
        {
            return NotFound();
        }
        return Ok(ReportResponse.FromEntity(report));
    }

    [HttpGet]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public async Task<IActionResult> ListReports(
        [FromQuery] string? userId = null,
        [FromQuery] ReportStatus? status = null)
    {
        List<Report> reports;
        if (userId != null)
        {
            reports = await _reportService.GetReportsByUserAsync(userId);
        }
        else if (status != null)
        {
            reports = await _reportService.GetReportsByStatusAsync(status.Value);
        }
        else
        {
            reports = await _reportService.GetReportsByStatusAsync(ReportStatus.COMPLETED);
        }

        var responses = reports.Select(ReportResponse.FromEntity).ToList();
        return Ok(new { reports = responses, total = responses.Count });
    }

    [HttpGet("{id:long}/download")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    public async Task<IActionResult> DownloadReport(long id)
    {
        var report = await _reportService.GetReportAsync(id);
        if (report == null)
        {
            return NotFound();
        }

        if (report.Status == ReportStatus.GENERATING || report.Status == ReportStatus.PENDING)
        {
            return Conflict();
        }

        if (report.Status == ReportStatus.FAILED || report.FilePath == null)
        {
            return NotFound();
        }

        if (!System.IO.File.Exists(report.FilePath))
        {
            _logger.LogWarning("Report file missing: {Path}", report.FilePath);
            return NotFound();
        }

        var fileContent = await System.IO.File.ReadAllBytesAsync(report.FilePath);
        var contentType = GetContentType(report.ReportType);
        var fileName = Path.GetFileName(report.FilePath);

        return File(fileContent, contentType, fileName);
    }

    [HttpDelete("{id:long}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> DeleteReport(long id)
    {
        var deleted = await _reportService.DeleteReportAsync(id);
        if (!deleted)
        {
            return NotFound();
        }
        return NoContent();
    }

    private static string GetContentType(ReportType reportType) => reportType switch
    {
        ReportType.PDF => "application/pdf",
        ReportType.CSV => "text/csv",
        ReportType.EXCEL => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        _ => "application/octet-stream"
    };
}
